_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,310 @@
+
+The recommendation here is to configure the Kafka cluster with `auto.create.topics.enable: false` in order to reduce the likelihood of hitting this problem.
+Unfortunately, disabling auto topic creation doesn't help with applications that use Kafka Admin client to directly create required topics (i.e. Kafka Streams apps).
+In this case, the application should be written to wait for topic existence, or use an init container to do the same.
+
+#### Finalizers
+
+[Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers) are used by default to ensure that topic deletion events are not missed when the UTO is not running.
+If this happens, you end up with the `KafkaTopic` resource deleted in Kubernetes by garbage collection, and a Kafka topic still present in Kafka.
+
+It is recommended to delete `KafkaTopic` resources when the operator is running in order to trigger the required cleanups.
+A common pitfall is that an attempt to delete a namespace or Custom Resource Definition becomes stuck in a "terminating" state.
+This happens when you try to delete everything at once and the operator does not have time to delete all the `KafkaTopic` resources.
+In such a scenario, you can remove the finalizers from all `KafkaTopic` resources simultaneously using the following command.
+
+```sh
+$ kubectl get kt -o yaml | yq 'del(.items[].metadata.finalizers[])' | kubectl apply -f -
+kafkatopic.kafka.strimzi.io/topic-3 configured
+kafkatopic.kafka.strimzi.io/topic-1 configured
+kafkatopic.kafka.strimzi.io/topic-2 configured
+```
+
+In development and test environment you can disable the use of finalizers with the following configuration.
+
+```sh
+# from Kafka custom resource
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+    template:
+      topicOperatorContainer:
+        env:
+          - name: STRIMZI_USE_FINALIZERS
+            value: "false"
+
+# from standalone Deployment
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: strimzi-topic-operator
+spec:
+  template:
+    spec:
+      containers:
+        - name: STRIMZI_USE_FINALIZERS
+          value: "false"
+```
+
+### Upgrading from BTO to UTO
+
+The UTO is available as alpha feature since Strimzi 0.36.0, so we need to enable a feature gate in order to use it.
+[Strimzi 0.38.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.38.0) adds Prometheus metrics and pause reconciliation feature.
+
+**The plan is to move it to beta in Strimzi 0.39.0, where it will be enabled by default.**
+**Tests show that UTO consumes approximately 40% less CPU and 30% less memory than the BTO.**
+
+```sh
+$ kubectl set env deploy strimzi-cluster-operator STRIMZI_FEATURE_GATES="+UnidirectionalTopicOperator"
+deployment.apps/strimzi-cluster-operator updated
```

Comment:
maybe move this up to sit under _so we need to enable a feature gate in order to use it._?

Reply:
I think it's better to warn users before the command.